keel-actuator/src/main/kotlin/com/netflix/spinnaker/keel/actuation/EnvironmentPromotionChecker.kt

```diff
@@ -67,8 +86,34 @@ class EnvironmentPromotionChecker(
       }
   }
 
-  private fun Environment.hasSupportedConstraint(constraintEvaluator: ConstraintEvaluator<*>) =
-    constraints.any { it.javaClass.isAssignableFrom(constraintEvaluator.constraintType) }
+  private fun checkConstraints(
+    evaluators: List<ConstraintEvaluator<*>>,
+    artifact: DeliveryArtifact,
+    deliveryConfig: DeliveryConfig,
+    version: String,
+    environment: Environment
+  ): Boolean {
+    if (evaluators.isEmpty()) {
+      return true
+    } else {
+      return evaluators.all { evaluator ->
+        val constraint = environment.constraintForEvaluator(evaluator)
+        if (constraint == null) {
```

Comment:
Sorry to intrude, but I wanted to see IIUC: constraint evaluators are just `@Components` wired into that list. An environment may or may not have a constraint matching each of those evaluators, hence it can be null. Right? 😃

Reply:
@luispollo is right, most environments aren't going to utilize every constraint type, so `constraintForEvaluator` will only return a constraint a minority of times. When it doesn't, it means the evaluator in question doesn't get a chance to veto, so it counts as an automatic `true` for `evaluators.all {}` instead of having its `canPromote()` called.

Reply:
Oh, I see.